src/compiler/transformers/declarations/emitBinder.ts

```diff
@@ -0,0 +1,914 @@
+        postBindingAction.forEach(fn => fn());
+
+        function createAnonymousEmitSymbol(node: Declaration): EmitDeclarationSymbol {
+            const symbol = createEmitSymbol();
+            symbol.declarations.push(node);
+            setValueDeclaration(symbol as Symbol, node);
+            node.symbol = symbol as Symbol;
+            return symbol;
+        }
+        function createEmitSymbol(): EmitDeclarationSymbol {
+            return {
+                parent: currentSymbol,
+                declarations: [],
+                flags: 0,
+            };
+        }
+        function getSymbol(table: EmitDeclarationSymbolTable, name: MemberKey) {
+            let symbol = table.get(name);
+            if (!symbol) {
+                symbol = createEmitSymbol();
+                symbol.name = name;
+                table.set(name, symbol);
+            }
+            return symbol;
+        }
+        function addLocalAndExportDeclaration(name: LocalAndExportName | MemberKey | undefined, node: Declaration, flags: SymbolRegistrationFlags, isExport: boolean, isBlockScoped = true) {
+            const { exportName, localName } = typeof name === "object" ? name : { exportName: name, localName: name };
+            if (isExport) {
+                const exportKind = flags.includes & SymbolFlags.Value ? SymbolFlags.ExportValue : 0;
+                const localSymbol = addLocalOnlyDeclaration(localName, node, { includes: exportKind, excludes: flags.excludes }, isBlockScoped);
+                const exportSymbol = addExportOnlyDeclaration(exportName, node, flags);
+                localSymbol.exportSymbol = exportSymbol;
+                // Export symbol can be undefined if the export modifier was placed in an unexpected position.
+                // We just assume the local symbol should be used. There are already bigger issues in the file anyway.
+                return exportSymbol ?? localSymbol;
+            }
+            else {
+                return addLocalOnlyDeclaration(localName, node, flags, isBlockScoped);
+            }
+        }
+        function addExportOnlyDeclaration(name: MemberKey | undefined, node: Declaration, flagsAndForbiddenFlags: SymbolRegistrationFlags) {
+            if (!currentExportsSymbolTable) {
+                return undefined;
+            }
+            return addDeclaration(currentExportsSymbolTable, name, node, flagsAndForbiddenFlags);
+        }
+        function addLocalOnlyDeclaration(name: MemberKey | undefined, node: Declaration, flagsAndForbiddenFlags: SymbolRegistrationFlags, isBlockScoped = true) {
+            return addDeclaration(isBlockScoped ? currentLocalSymbolTable : currentFunctionLocalSymbolTable, name, node, flagsAndForbiddenFlags);
+        }
+
+        function addDeclaration(table: EmitDeclarationSymbolTable, name: MemberKey | undefined, node: Declaration, { includes, excludes }: SymbolRegistrationFlags) {

```

Comment:
there's a lot of repeated code here and in createAnonymousEmitSymbol/getSymbol. Comparing those with this raises quite a few questions. Probably it would be better to de-dupe the code, making it all behave the same way.